feat: add jobs

atrauzzi opened the pull request:
# Description

Added initial support for jobs.

## Issue reference

Closes: https://github.com/dapr/js-sdk/issues/667

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [ ] Code compiles correctly
- [ ] Created/updated tests
- [ ] Extended the documentation

atrauzzi committed:
Rough-in the contracts.

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

atrauzzi committed:
Implement registration.

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

atrauzzi committed:
Implement the rest of the jobs utility methods.

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

atrauzzi committed:
Cleanup.

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

atrauzzi committed:
Cleanup.

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

WhitWaldo commented:
@atrauzzi This runs a linter (for now) as part of the build pipeline and failures will stop the build. Would you mind fixing the lint errors locally and pushing a commit with those changes as well so we can run the checks?

atrauzzi committed:
Temporarily disable some eslint checks for builds.

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

WhitWaldo commented on src/interfaces/Client/IClientJobs.ts:
```diff
@@ -0,0 +1,31 @@
+    ttl: string | null,
+  ): Promise<void>;
+
+  get(jobName: string): Promise<Job>;
```
I believe all jobs will return with a payload even if it's empty. Worth testing and double checking.

WhitWaldo replied:
The [relevant proto](https://github.com/dapr/dapr/blob/2e7c61e933099b9d40fcdefbd57cc3e81069915a/dapr/proto/runtime/v1/dapr.proto#L1267) does not mark the data as optional

WhitWaldo commented on src/interfaces/Server/IServerJobs.ts:
```diff
@@ -0,0 +1,20 @@
+import { TypeDaprJobsCallback } from "../../types/DaprJobsCallback.type";
+
+export default interface IServerJobs {
+  listen(jobName: string, callback: TypeDaprJobsCallback): void;
```
Same here - I don't know that the server will ever pass along a job invocation without a payload (e.g. though of course it might be empty).

WhitWaldo commented on src/types/jobs/JobSchedule.type.ts:
```diff
@@ -0,0 +1,29 @@
+
+type EveryPeriod = `@every ${string}`;
+
+// note: This can get crazy, more than TS can really handle.
```
You might consider taking a look at [how I built this out](https://github.com/dapr/dotnet-sdk/blob/master/src/Dapr.Jobs/Models/DaprJobSchedule.cs) in the .NET SDK to accommodate each variation.

WhitWaldo replied:
Most importantly - there are different sets of arguments that need to be provided based on this schedule and the other arguments used.

WhitWaldo reviewed (changes requested):
Excellent start - still a bit to go, but it's a good foundation.

atrauzzi committed:
Mimic the cron expression from the dotnet library.

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

atrauzzi committed:
Rename each to every.

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

WhitWaldo commented on src/types/jobs/CronExpression.type.ts:
```diff
@@ -0,0 +1,257 @@
+  Month,
+}
+
+export type TimeValue = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20 | 21 | 22 | 23 | 24 | 25 | 26 | 27 | 28 | 29 | 30 | 31 | 32 | 33 | 34 | 35 | 36 | 37 | 38 | 39 | 40 | 41 | 42 | 43 | 44 | 45 | 46 | 47 | 48 | 49 | 50 | 51 | 52 | 53 | 54 | 55 | 56 | 57 | 58 | 59;
```
I'm sure this was fun :)

WhitWaldo commented on src/types/jobs/CronExpression.type.ts:
```diff
@@ -0,0 +1,257 @@
+    /(([0-1]?\d)|(2[0-3])-([0-1]?\d)|(2[0-3]))|(([0-1]?\d)|(2[0-3]),?)|(\*(\/([0-1]?\d)|(2[0-3]))?)/;
+  private static readonly DayOfMonthRegexText =
+    /\*|(\*\/(([0-2]?\d)|(3[0-1])))|(((([0-2]?\d)|(3[0-1]))(-(([0-2]?\d)|(3[0-1])))?))/;
+  private static readonly MonthRegexText =
```
Does JS regex support lookback syntax? This might need a tweak.

WhitWaldo reviewed (changes requested):
Coming along - note about the regexps though

atrauzzi committed:
Expanding things.

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

atrauzzi committed:
Change case of static method.

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

atrauzzi committed:
Cleaning up, getting ready for some regex-fu.

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

atrauzzi committed:
Crons make me sad.

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

atrauzzi committed:
All crons and regex...

(...make Jack a dull boy)

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

atrauzzi committed:
Exploring an e2e approach.

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

WhitWaldo commented on src/interfaces/Client/IClientJobs.ts:
```diff
@@ -11,17 +11,17 @@ See the License for the specific language governing permissions and
 */
 
-import { JobSchedule } from "../../types/jobs/JobSchedule.type";
+import { Schedule } from "../../types/jobs/JobSchedule.type";
```
Especially if the file only contains a single type, the type name should match the file name.

WhitWaldo commented on src/interfaces/Server/IServerJobs.ts:
```diff
@@ -11,10 +11,7 @@ See the License for the specific language governing permissions and
-  listen(jobName: string, callback: TypeDaprJobsCallback): void;
-
-  listen<DataType>(jobName: string, callback: TypeDaprJobsCallback<DataType>): void;
+  listen(jobName: string, callback: (data: any) => unknown): void;
```
Why does this return an `unknown`? While the callback needn't necessarily return anything, the SDK does need to know that it was handled without error so it can send back a 200 OK.

WhitWaldo commented on test/e2e/jobs/jobs.test.ts:
```diff
@@ -11,18 +11,139 @@ See the License for the specific language governing permissions and
+
+        const job = await client?.jobs.get("test");
+
+        await new Promise(resolve => setTimeout(resolve, 5000));
```
The scheduler does not guarantee strict and immediate invocation, just that it'll be near the trigger point. Setting a timeout to 5000 ms suggests it's unlikely you'll see this invoked more than 4 times as there won't necessarily be strictly 1000 ms between each invocation.

I'd put the timeout at 6 seconds and measure for 5 invocations. Also, limit the repeats to 5 so the job cleans itself up.

WhitWaldo reviewed (changes requested):
More thoughts

atrauzzi committed:
Cleaning.

 - Addresses
   - https://github.com/dapr/js-sdk/pull/688#discussion_r2008844993
   - https://github.com/dapr/js-sdk/pull/688#discussion_r2008846881

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

WhitWaldo committed:
Remove prettifier from build (#684)

* Removing prettifier from the build pipeline as it's an unnecessary step during a build operation.
* Version number updated in package-lock.json

Signed-off-by: Whit Waldo <whit.waldo@innovian.net>
Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

ConstantinChirila commented on src/types/jobs/CronExpression.type.ts:
```diff
@@ -0,0 +1,355 @@
+  Month,
+}
+
+export type SecondOrMinute = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20 | 21 | 22 | 23 | 24 | 25 | 26 | 27 | 28 | 29 | 30 | 31 | 32 | 33 | 34 | 35 | 36 | 37 | 38 | 39 | 40 | 41 | 42 | 43 | 44 | 45 | 46 | 47 | 48 | 49 | 50 | 51 | 52 | 53 | 54 | 55 | 56 | 57 | 58 | 59;
+export type Hour = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20 | 21 | 22 | 23;
+export type DayOfMonth = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20 | 21 | 22 | 23 | 24 | 25 | 26 | 27 | 28 | 29 | 30 | 31;
+export type DayOfWeekNumber = 0 | 1 | 2 | 3 | 4 | 5 | 6;
+export type MonthNumber = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;
```
This is a bit of a NIT, but theres a different way of adding type for these situations, which ive used before.
You would have first a type that would help you build such types: 

```
type BuildNumberRange<T extends number, U extends number, R extends number[] = []> = R['length'] extends U
  ? R[number]
  : BuildNumberRange<T, U, [...R, R['length'] & number]>;
```

And then you can have your types like this: 
```
export type SecondOrMinute = BuildNumberRange<0, 59>
export type Hour  = BuildNumberRange<0, 23>
export type DayOfMonth = BuildNumberRange<1, 31>
export type DayOfWeekNumber = BuildNumberRange<0, 6>
export type MonthNumber = BuildNumberRange<1, 12>
```
Just an idea. I realise the `BuildNumberRange` type is quite hard to reason with for new comers, but its the type of "set it and forget it".

WhitWaldo committed:
Removed references to Workflow components (#687)

Removed references to Workflow components and replaced with "dapr" for the name in the HTTP client.

Signed-off-by: Whit Waldo <whit.waldo@innovian.net>
Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

WhitWaldo committed:
Removes stale bot from repo (#690)

Removes stale bot + adds me to repo owners for bot actions

Signed-off-by: Whit Waldo <whit.waldo@innovian.net>
Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

joebowbeer commented:
Error

```
== APP ==     Error: 14 UNAVAILABLE: connection error: desc = "transport: Error while dialing: dial tcp 127.0.0.1:50001: connect: connection refused"
== APP ==         at callErrorFromStatus (/home/runner/work/js-sdk/js-sdk/node_modules/@grpc/grpc-js/src/call.ts:82:17)
== APP ==         at Object.onReceiveStatus (/home/runner/work/js-sdk/js-sdk/node_modules/@grpc/grpc-js/src/client.ts:360:55)
== APP ==         at Object.onReceiveStatus (/home/runner/work/js-sdk/js-sdk/node_modules/@grpc/grpc-js/src/client-interceptors.ts:458:34)
== APP ==         at Object.onReceiveStatus (/home/runner/work/js-sdk/js-sdk/node_modules/@grpc/grpc-js/src/client-interceptors.ts:419:48)
== APP ==         at /home/runner/work/js-sdk/js-sdk/node_modules/@grpc/grpc-js/src/resolving-call.ts:163:24
== APP ==         at processTicksAndRejections (node:internal/process/task_queues:78:11)
== APP ==     for call at
== APP ==         at ServiceClientImpl.makeUnaryRequest (/home/runner/work/js-sdk/js-sdk/node_modules/@grpc/grpc-js/src/client.ts:325:42)
== APP ==         at ServiceClientImpl.invokeService (/home/runner/work/js-sdk/js-sdk/node_modules/@grpc/grpc-js/src/make-client.ts:189:15)
== APP ==         at /home/runner/work/js-sdk/js-sdk/src/implementation/Client/GRPCClient/invoker.ts:64:14
== APP ==         at new Promise (<anonymous>)
== APP ==         at GRPCClientInvoker.invoke (/home/runner/work/js-sdk/js-sdk/src/implementation/Client/GRPCClient/invoker.ts:63:12)
== APP ==         at Object.<anonymous> (/home/runner/work/js-sdk/js-sdk/test/e2e/grpc/server.test.ts:78:21) {
== APP ==       code: 14,
== APP ==       details: 'connection error: desc = "transport: Error while dialing: dial tcp 127.0.0.1:50001: connect: connection refused"',
== APP ==       metadata: Metadata {
== APP ==         internalRepr: Map(2) { 'traceparent' => [Array], 'grpc-trace-bin' => [Array] },
== APP ==         options: {}
== APP ==       }
== APP ==     }
== APP == 
== APP ==       at Object.<anonymous> (test/e2e/grpc/server.test.ts:81:17)
```

WhitWaldo commented:
> ```
> == APP ==     Error: 14 UNAVAILABLE: connection error: desc = "transport: Error while dialing: dial tcp 127.0.0.1:50001: connect: connection refused"
> == APP ==         at callErrorFromStatus (/home/runner/work/js-sdk/js-sdk/node_modules/@grpc/grpc-js/src/call.ts:82:17)
> == APP ==         at Object.onReceiveStatus (/home/runner/work/js-sdk/js-sdk/node_modules/@grpc/grpc-js/src/client.ts:360:55)
> == APP ==         at Object.onReceiveStatus (/home/runner/work/js-sdk/js-sdk/node_modules/@grpc/grpc-js/src/client-interceptors.ts:458:34)
> == APP ==         at Object.onReceiveStatus (/home/runner/work/js-sdk/js-sdk/node_modules/@grpc/grpc-js/src/client-interceptors.ts:419:48)
> == APP ==         at /home/runner/work/js-sdk/js-sdk/node_modules/@grpc/grpc-js/src/resolving-call.ts:163:24
> == APP ==         at processTicksAndRejections (node:internal/process/task_queues:78:11)
> == APP ==     for call at
> == APP ==         at ServiceClientImpl.makeUnaryRequest (/home/runner/work/js-sdk/js-sdk/node_modules/@grpc/grpc-js/src/client.ts:325:42)
> == APP ==         at ServiceClientImpl.invokeService (/home/runner/work/js-sdk/js-sdk/node_modules/@grpc/grpc-js/src/make-client.ts:189:15)
> == APP ==         at /home/runner/work/js-sdk/js-sdk/src/implementation/Client/GRPCClient/invoker.ts:64:14
> == APP ==         at new Promise (<anonymous>)
> == APP ==         at GRPCClientInvoker.invoke (/home/runner/work/js-sdk/js-sdk/src/implementation/Client/GRPCClient/invoker.ts:63:12)
> == APP ==         at Object.<anonymous> (/home/runner/work/js-sdk/js-sdk/test/e2e/grpc/server.test.ts:78:21) {
> == APP ==       code: 14,
> == APP ==       details: 'connection error: desc = "transport: Error while dialing: dial tcp 127.0.0.1:50001: connect: connection refused"',
> == APP ==       metadata: Metadata {
> == APP ==         internalRepr: Map(2) { 'traceparent' => [Array], 'grpc-trace-bin' => [Array] },
> == APP ==         options: {}
> == APP ==       }
> == APP ==     }
> == APP == 
> == APP ==       at Object.<anonymous> (test/e2e/grpc/server.test.ts:81:17)
> ```

What's this in reference to?

joebowbeer commented:
> What's this in reference to?

@WhitWaldo This is the failure from the test run of the previous commit.

However, I don't think it sheds much light. I think the interesting failure is in the scheduler's (Testcontainer) log:

```
"failed to invoke schedule app job: error returned from app channel while sending triggered job to app: Post \"http://127.0.0.1:8070/job/test\": EOF"
```

joebowbeer commented on test/e2e/jobs/jobs.test.ts:
```diff
@@ -0,0 +1,149 @@
+    beforeAll(async () => {
+
+        daprScheduler = await new GenericContainer("ghcr.io/dapr/dapr")
+          .withExposedPorts(
```
I'm looking at this to see how a real DaprContainer Testcontainer would be used.

Note: fixed host port bindings are not recommended

https://node.testcontainers.org/features/containers/#exposing-container-ports

Consider using a network alias

https://node.testcontainers.org/features/networking/#network-aliases

joebowbeer replied:
What specifically needs to communicate with what?

Checkout the use of socat:

https://node.testcontainers.org/features/containers/#socatcontainer-as-a-tcp-proxy

Test:

https://github.com/testcontainers/testcontainers-node/blob/main/packages/testcontainers/src/socat/socat-container.test.ts

joebowbeer commented on test/e2e/jobs/jobs.test.ts:
```diff
@@ -0,0 +1,149 @@
+    beforeEach(async () => {
+
+      server = new DaprServer({
+        serverHost: "127.0.0.1",
```
Is this right? Or should this be the network alias or `host.testcontainers.internal`?

https://node.testcontainers.org/features/networking/#expose-host-ports-to-container

joebowbeer replied:
You're running docker-in-docker in CI, so I doubt that will resolve to the machine running the tests.

My advice:
1. This should run in a custom network. I think this is essential.
2. This should not use fixed host ports (because it's good practice)

By the way, I'm familiar with the following tests but there might be better examples in other modules:

https://github.com/testcontainers/testcontainers-node/blob/main/packages/modules/localstack/src/localstack-container.test.ts

https://github.com/testcontainers/testcontainers-node/blob/main/packages/modules/k3s/src/k3s-container.test.ts

atrauzzi committed:
WIP

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

atrauzzi committed:
Apply: https://github.com/dapr/js-sdk/pull/688/files#r2008853805

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

joebowbeer commented on test/e2e/jobs/jobs.test.ts:
```diff
@@ -0,0 +1,162 @@
+          .withCommand([
+            "./daprd",
+            "--app-id", "dapr-js-sdk-testing",
+            // todo: Need to figure out how to tell daprd where my app can be found as it's not on `localhost`
+            // "--app-endpoint", "host.testcontainers.internal",
```
From Java's DaprContainer, it looks like there's a CMD for that?

https://github.com/dapr/java-sdk/blob/8b8684a2db481c704b9aa7e0fbf5198d8ca42e10/testcontainers-dapr/src/main/java/io/dapr/testcontainers/DaprContainer.java#L289

atrauzzi committed:
WIP

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

joebowbeer commented on test/e2e/jobs/jobs.test.ts:
```diff
@@ -0,0 +1,166 @@
+            "--app-port", "8070",
+            "--dapr-grpc-port", "8081",
+            "--dapr-http-port", "8082",
+            "--scheduler-host-address", `scheduler:${getPort(daprScheduler, 8083)}`,
```
If they are both running on the same network, why use the host port?

```suggestion
            "--scheduler-host-address", `scheduler:8083`,
```

atrauzzi committed:
WIP

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

joebowbeer commented on test/e2e/jobs/jobs.test.ts:
```diff
@@ -0,0 +1,154 @@
+import { CommunicationProtocolEnum, DaprClient, DaprServer } from "../../../src";
+// import { AbstractWaitStrategy } from "testcontainers/build/wait-strategies/wait-strategy";
+
+jest.setTimeout(40 * 1000);
```
When working with testcontainers these timeouts are usually much longer, e.g., 120_000 or 180_000.

The testcontainers default wait is 60_000, so you want jest to wait at least that long.

joebowbeer commented on test/e2e/jobs/jobs.test.ts:
```diff
@@ -0,0 +1,154 @@
+
+    beforeAll(async () => {
+
+      await TestContainers.exposeHostPorts(8070);
```
In all the examples, this is called after there is a server listening on this port.

Try moving this after the 8070 server is started.

joebowbeer commented on test/e2e/jobs/jobs.test.ts:
```diff
@@ -0,0 +1,154 @@
+        .withTmpFs({
+          "/data": "rw",
+        })
+        // note: Because dapr containers don't have `sh` or `bash` inside, this is kind of the best health check.
```
Two comments:

1. Wait.forHttp(path, port) works for me, and is what Dapr's Testcontainer for Java does:

https://github.com/dapr/java-sdk/blob/ab8e41111de17ea5bc47b9d94dbab6fe30577590/testcontainers-dapr/src/main/java/io/dapr/testcontainers/DaprContainer.java#L60

2. The timeout should be much longer, e.g., 120_000

joebowbeer replied:
I suggest something like the following for the wait strategy

```ts
  .withWaitStrategy(Wait.forHttp("/v1.0/healthz/outbound", DAPRD_HTTP_PORT)
    .forStatusCodeMatching((statusCode) => statusCode >= 200 && statusCode <= 399))
  .withStartupTimeout(120_000)
```

joebowbeer commented on test/e2e/jobs/jobs.test.ts:
```diff
@@ -0,0 +1,154 @@
+      console.info(`Scheduler: ${getIp(daprScheduler)}`);
+      console.info(`Daemon: ${getIp(daprd)}`);
+
+      server = new DaprServer({
```
Do you need DaprServer or can you just use an express server?

I think there's a chicken-egg problem with DaprServer in these tests because exposeHostPorts should be called after the server starts but before starting daprd.

joebowbeer committed:
chore: update dependencies (#699)

Signed-off-by: Joe Bowbeer <joe.bowbeer@gmail.com>
Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

atrauzzi committed:
Triggers once.

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

atrauzzi committed:
WIP

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

atrauzzi committed:
A few hours and some caffeine later...  We got it!

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

atrauzzi committed:
Clean up.

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

atrauzzi committed:
Merge branch 'main' into add-jobs

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

# Conflicts:
#	package-lock.json

atrauzzi committed:
Add test for removing a job.

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

atrauzzi committed:
Merge branch 'main' into add-jobs

atrauzzi committed:
Cleanup and stub some basic docs.

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

WhitWaldo reviewed (commented):
Just want to make sure we've got strong test coverage - reviewed on GitHub, so I can't run a check myself, but if it looks good to you, I'm satisifed.

Also, please remove the gRPC implementations of the Jobs bits. The way I see it, the SDK need only implement one or the other and it's up to the runtime to ensure that there's parity in the APIs. We have no need to expose both and have the developer pick one as that would just be confusing. ~~Rather, you've already got HTTP in place, so let's delete the gRPC side.~~

I flipped these. I'd prefer to keep gRPC and drop HTTP.

WhitWaldo committed:
Merge branch 'main' into add-jobs

WhitWaldo commented:
@atrauzzi Do you have any interest in getting this over the finish line?

atrauzzi committed:
Fix up and add tests for prepare day of week value list.

Signed-off-by: Alexander Trauzzi <acj@trauzzi.me>

WhitWaldo commented:
I'm good with just implementing the HTTP bits (already done) and just leaving gRPC unimplemented. We can call this out in the documentation that this API is currently only supported when using it in an HTTP context and throw an exception that reflects the same. 

Protos should be updated per the recently merged conversation API, so if you can merge with that, I think this should generally be good to go, right?

WhitWaldo commented:
@atrauzzi Do you have any bandwidth to get this over the finish line for 1.18? If not, that's perfectly fine, and I'd be happy to pick it up and cite you in the release notes. Do note that the Jobs API is going stable this release, so you'll need to update your type/method names accordingly.

atrauzzi commented:
Apologies, I **definitely** didn't over the last 2 weeks :sweat_smile:, tough to guarantee my availability.

Aside from the proto changes, any other fixups?  May be able to this week if there are no surprises...
